File: source/state_representation/test/tests/space/test_spatial_state.cpp
#include <gtest/gtest.h>

#include "state_representation/space/SpatialState.hpp"

using namespace state_representation;

TEST(SpatialStateTest, Constructors) {
  SpatialState state0;
  EXPECT_EQ(state0.get_type(), StateType::SPATIAL_STATE);
  EXPECT_EQ(state0.get_name(), "");
  EXPECT_EQ(state0.get_reference_frame(), "world");
  EXPECT_TRUE(state0.is_empty());
}

TEST(SpatialStateTest, Compatibility) {
  SpatialState state1("test", "robot");
  EXPECT_TRUE(state1.is_incompatible(SpatialState("state")));
  EXPECT_TRUE(state1.is_incompatible(SpatialState("test")));
  EXPECT_FALSE(state1.is_incompatible(SpatialState("state", "robot")));
  EXPECT_FALSE(state1.is_incompatible(SpatialState("state", "test")));
  EXPECT_FALSE(state1.is_incompatible(SpatialState("robot")));
}

TEST(SpatialStateTest, Swap) {
  SpatialState state1("cartesian");
  SpatialState state2("joint", "robot");
  swap(state1, state2);
  EXPECT_EQ(state1.get_name(), "joint");
  EXPECT_EQ(state1.get_reference_frame(), "robot");
  EXPECT_TRUE(state1.is_empty());

  EXPECT_EQ(state2.get_name(), "cartesian");
  EXPECT_EQ(state2.get_reference_frame(), "world");
  EXPECT_TRUE(state2.is_empty());
}

File: python/test/state_representation/space/test_spatial_state.py
import unittest
import copy

from state_representation import SpatialState, StateType
from ..test_state import STATE_METHOD_EXPECTS

SPATIAL_STATE_METHOD_EXPECTS = [
    'get_reference_frame',
    'set_reference_frame'
]


class TestState(unittest.TestCase):

    def test_callable_methods(self):
        methods = [m for m in dir(SpatialState) if callable(getattr(SpatialState, m))]
        for expected in STATE_METHOD_EXPECTS:
            self.assertIn(expected, methods)
        for expected in SPATIAL_STATE_METHOD_EXPECTS:
            self.assertIn(expected, methods)

    def test_constructors(self):
        state1 = SpatialState()
        self.assertEqual(state1.get_type(), StateType.SPATIAL_STATE)
        self.assertEqual(state1.get_name(), "")
        self.assertEqual(state1.get_reference_frame(), "world")
        self.assertTrue(state1.is_empty())

        state2 = SpatialState("test", "robot")
        self.assertEqual(state2.get_type(), StateType.SPATIAL_STATE)
        self.assertEqual(state2.get_name(), "test")
        self.assertEqual(state2.get_reference_frame(), "robot")

        state3 = SpatialState(state2)
        self.assertEqual(state2.get_type(), state3.get_type())
        self.assertEqual(state2.get_name(), state3.get_name())
        self.assertEqual(state2.get_reference_frame(), state3.get_reference_frame())
        self.assertEqual(state2.is_empty(), state3.is_empty())

    def test_copy(self):
        state = SpatialState("test", "robot")
        for state_copy in [copy.copy(state), copy.deepcopy(state)]:
            self.assertEqual(state.get_type(), state_copy.get_type())
            self.assertEqual(state.get_name(), state_copy.get_name())
            self.assertEqual(state.get_reference_frame(), state_copy.get_reference_frame())
            self.assertEqual(state.is_empty(), state_copy.is_empty())

    def test_compatibility(self):
        state1 = SpatialState("test", "robot")
        self.assertTrue(state1.is_incompatible(SpatialState("state")))
        self.assertTrue(state1.is_incompatible(SpatialState("test")))
        self.assertFalse(state1.is_incompatible(SpatialState("state", "robot")))
        self.assertFalse(state1.is_incompatible(SpatialState("state", "test")))
        self.assertFalse(state1.is_incompatible(SpatialState("robot")))


if __name__ == '__main__':
    unittest.main()
